velox/functions/prestosql/ArraySum.cpp

```diff
@@ -0,0 +1,159 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/ComparatorUtil.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+///
+/// Implements the array_sum function.
+/// See documentation at https://prestodb.io/docs/current/functions/array.html
+///
+template <typename IT, typename OT>
+class ArraySumFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args[0]->as<ArrayVector>();
+    VELOX_CHECK(arrayVector);
+    auto elementsVector = arrayVector->elements();
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+    vector_size_t numRows = arrayVector->size();
+
+    // Prepare result vector for writing
+    BaseVector::ensureWritable(rows, outputType, context->pool(), result);
+    auto resultValues = (*result)->template asFlatVector<OT>();
+
+    rows.template applyToSelected([&](vector_size_t row) {
+      if (arrayVector->isNullAt(row)) {
+        resultValues->setNull(row, true);
+      } else {
+        int start = arrayVector->offsetAt(row);
+        int end = start + arrayVector->sizeAt(row);
+
+        OT sum = 0;
+        for (; start < end; start++) {
+          if (!elements->isNullAt(start)) {
+            sum += elements->template valueAt<IT>(start);
+          }
+        }
+        resultValues->set(row, sum);
+      }
+    });
+  }
+};
+
+// Validate number of parameters and types.
+void validateType(const std::vector<exec::VectorFunctionArg>& inputArgs) {
+  VELOX_USER_CHECK_EQ(
+      inputArgs.size(), 1, "array_sum requires exactly one parameter");
+
+  auto arrayType = inputArgs.front().type;
+  VELOX_USER_CHECK_EQ(
+      arrayType->kind(),
+      TypeKind::ARRAY,
+      "array_sum requires argument of type ARRAY");
+
+  auto valueTypeKind = inputArgs.front().type->childAt(0)->kind();
+  bool isCoercibleToDouble = valueTypeKind == TypeKind::TINYINT ||
+      valueTypeKind == TypeKind::SMALLINT ||
+      valueTypeKind == TypeKind::INTEGER || valueTypeKind == TypeKind::BIGINT ||
```

Comment:
can we compress this into `valueTypeKind >= TypeKind::TINYINT && valueTypeKind <= TypeKind::DOUBLE`

Reply:
I think that it's possible to do, but it would make it less clear what we are doing. I think it should stay this way.